numpy/random/mtrand/mtrand.pyx

```diff
@@ -1426,15 +1435,272 @@ cdef class RandomState:
+
+        Return random np.int32 integers between `low` and `high`, inclusive.
+
+        Return random integers from the "discrete uniform" distribution in the
+        closed interval [`low`, `high`].  If `high` is None (the default),
+        then results are from [1, `low`].
+
+        Parameters
+        ----------
+        low : int
+            Lowest (signed) integer to be drawn from the distribution (unless
+            ``high=None``, in which case this parameter is the *highest* such
+            integer).
+        high : int, optional
+            If provided, the largest (signed) integer to be drawn from the
+            distribution (see above for behavior if ``high=None``).
+        size : int or tuple of ints, optional
+            Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
+            ``m * n * k`` samples are drawn.  Default is None, in which case a
+            single value is returned.
+
+        Returns
+        -------
+        out : int or ndarray of npy_intp
+              `size`-shaped array of random integers from the appropriate
+              distribution, or a single such random int if `size` not provided.
+
+        See Also
+        --------
+        random.random_integers : Similar function returning np.int type.
+        random.random_int64 : Similar function returning np.int64 type.
+        random.random_intp : Similar function returning np.intp type.
+
+        Notes
+        -----
+
+        Examples
+        --------
+        >>> np.random.random_integers(5)
+        4
+        >>> type(np.random.random_integers(5))
+        <type 'int'>
+        >>> np.random.random_integers(5, size=(3.,2.))
+        array([[5, 4],
+               [3, 3],
+               [4, 5]])
+
+        Choose five random numbers from the set of five evenly-spaced
+        numbers between 0 and 2.5, inclusive (*i.e.*, from the set
+        :math:`{0, 5/8, 10/8, 15/8, 20/8}`):
+
+        >>> 2.5 * (np.random.random_integers(5, size=(5,)) - 1) / 4.
+        array([ 0.625,  1.25 ,  0.625,  0.625,  2.5  ])
+
+        Roll two six sided dice 1000 times and sum the results:
+
+        >>> d1 = np.random.random_integers(1, 6, 1000)
+        >>> d2 = np.random.random_integers(1, 6, 1000)
+        >>> dsums = d1 + d2
+
+        Display results as a histogram:
+
+        >>> import matplotlib.pyplot as plt
+        >>> count, bins, ignored = plt.hist(dsums, 11, normed=True)
+        >>> plt.show()
+
+        """
+        cdef npy_int32 lo, hi
+        cdef npy_uint32 off, rng, buf
+        cdef npy_uint32 *out
+        cdef ndarray array "arrayObject"
+        cdef npy_intp cnt
+
         if high is not None and low > high:
             raise ValueError("low > high")
 
-        cdef long lo, hi, rv
-        cdef unsigned long diff
-        cdef long *array_data
+        if high is None:
+            lo = 1
+            hi = low
+        else:
+            lo = low
+            hi = high
+
```

Comment:
I wonder if we could wrap this ```if...else``` block into a ```try...except``` block to provide a useful message about ```OverflowError``` issues that we are seeing in #6812, something like:
```
try:
... # if-else block
except OverflowError:
raise ValueError("Boundary values too large for int32.  Please use random_int64 instead.")
```